### PR TITLE
[REVIEW] fetch thrust/cub from github

### PR DESCRIPTION
**Pull request by rongou**
Use `FetchContent` to obtain thrust/cub release 1.9.0. This is needed to get per-thread default stream to work properly. Otherwise the `rmm::device_vector` falls back to the legacy default stream for some operations.

@harrism @jrhemstad @kkraus14

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
@mt-jones can you review this?

**Comment by jrhemstad** [comment]
> Otherwise the rmm::device_vector falls back to the legacy default stream for some operations.

Users can build RMM against a custom Thrust version if they want to get this fix in the meantime

**Comment by rongou** [comment]
I agree this can potentially be a big change for cuDF, but I'm not so sure about RMM. The thrust api surface for RMM is pretty small and looks to be stable. Since RMM and cuDF are in two separate repos, this necessarily has to be a two-step process. I can create a PR under cuDF to also pull in the thrust/cub release and let it run through the CI. Would that help?

This is really the first step towards enabling PTDS for Spark. We probably should add some multi-threaded host tests for both RMM and cuDF to verify PTDS behavior. Changing the Spark build with the new thrust release is fine for a prototype, but it's hardly "production ready" when neither RMM or cuDF has been built or tested with the release.

**Comment by jrhemstad** [comment]
> I can create a PR under cuDF to also pull in the thrust/cub release and let it run through the CI. Would that help?

It would be a necessary first step, yes. But that alone isn't enough. We need to decide if we are okay with overriding Thrust that is shipped with the toolkit.

File: thirdparty/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,23 @@ endif()
...
+    cub
+    GIT_REPOSITORY https://github.com/thrust/cub
+    GIT_TAG        1.9.10
+    GIT_SHALLOW    true
+)
+
+FetchContent_GetProperties(cub)
+if(NOT cub_POPULATED)
+  FetchContent_Populate(cub)
+  # We are not using the cub CMake targets, so no need to call `add_subdirectory()`.
```
Comment (mt-jones) [inline thirdparty/CMakeLists.txt:34]: Is this comment necessary?

Comment (rongou) [reply thirdparty/CMakeLists.txt:34]: In the doc for `FetchContent` it shows a common pattern of populating and then adding subdirectory. I guess this is more of a reminder for ourselves why we are not doing that.

**Comment by mt-jones** [comment]
@jrhemstad is there a place where discussing thrust override is taking place?

**Comment by harrism** [comment]
@mt-jones the discussion was on slack, can't link it here.

**Comment by kkraus14** [comment]
> @mt-jones the discussion was on slack, can't link it here.

Sounds like we need to open a github issue with a summary of the discussion and current thoughts so that we can have an open conversation.

**Comment by leofang** [comment]
> Sounds like we need to open a github issue with a summary of the discussion and current thoughts so that we can have an open conversation.

I happened to open one here as you typed! 😂 
https://github.com/conda-forge/cub-feedstock/issues/4

**Comment by harrism** [comment]
> > Sounds like we need to open a github issue with a summary of the discussion and current thoughts so that we can have an open conversation.
> 
> @kkraus14 I happened to open one here as you typed! joy
> [conda-forge/cub-feedstock#4](https://github.com/conda-forge/cub-feedstock/issues/4)

@leofang this discussion is about RAPIDS depending on CUB and Thrust from github rather than from the CUDA toolkit, so discussing it on conda-forge would not be appropriate. :)  I opened a cuDF issue here: https://github.com/rapidsai/cudf/issues/5346

**Comment by harrism** [comment]
Merged https://github.com/rapidsai/cudf/pull/5315, so merging this.

**Pull request merged**